File: services/api/app/services/fixtures.py
from datetime import datetime, timedelta, timezone
from random import Random

from app.core.room_config import ROOM_GROUPS, SENSORS_PER_ROOM, expected_sensor_mappings
from app.services.normalizer import normalize_record
from app.services.status import room_status


def fixture_record(seed: int = 42) -> dict[str, object]:
    rng = Random(seed)
    record: dict[str, object] = {"LogTime": datetime.now(timezone.utc).isoformat()}
    for mapping in expected_sensor_mappings():
        base = 4.1 + rng.uniform(-1.4, 1.4)
        if mapping.room_number in {3, 5, 12, 14} and mapping.position_in_room in {1, 2, 3, 4, 5}:
            base = 7.1 + rng.uniform(0.0, 1.4)
        if mapping.room_number == 4 and mapping.position_in_room == 4:
            base = 5.8
        if mapping.room_number == 18:
            continue
        record[mapping.source_tag] = round(base, 1)
    return record


def dashboard_fixture() -> dict:
    normalized = normalize_record(fixture_record())
    readings_by_room: dict[int, list[dict]] = {}
    for reading in normalized["readings"]:
        readings_by_room.setdefault(reading["room_number"], []).append(reading)

    groups = []
    all_good_values = []
    alert_count = 0
    warning_count = 0
    critical_count = 0
    faults = 0

    for group_code, rooms in ROOM_GROUPS:
        room_items = []
        for room_number in rooms:
            sensors = readings_by_room[room_number]
            values = [sensor["value_c"] for sensor in sensors if sensor["quality"] == "good" and sensor["value_c"] is not None]
            status = room_status([sensor["status"] for sensor in sensors])
            all_good_values.extend(values)
            faults += sum(1 for sensor in sensors if sensor["status"] == "fault")
            critical_count += 1 if status == "critical" else 0
            warning_count += 1 if status == "warning" else 0
            alert_count += 1 if status in {"critical", "warning"} else 0
            room_items.append(
                {
                    "room_number": room_number,
                    "room_name": f"Cold Room {room_number}",
                    "group_code": group_code,
                    "average_c": round(sum(values) / len(values), 1) if values else None,
                    "status": status,
                    "sensors": sensors,
                    "active_alerts": [f"{status.title()} temperature condition"] if status in {"critical", "warning"} else [],
                }
            )
        groups.append({"group_code": group_code, "room_numbers": list(rooms), "rooms": room_items})

    return {
        "overview": {
            "total_rooms": 18,
            "total_sensors": SENSORS_PER_ROOM * 18,
            "average_c": round(sum(all_good_values) / len(all_good_values), 1) if all_good_values else None,
            "active_alerts": alert_count,
            "critical_alerts": critical_count,
            "warning_alerts": warning_count,
            "sensor_faults": faults,
            "last_log_time": normalized["logged_at"],
        },
        "groups": groups,
    }


def room_history_fixture(room_number: int) -> list[dict]:
    now = datetime.now(timezone.utc)
    points = []
    for index in range(24):
        snapshot = dashboard_fixture()
        room = next(room for group in snapshot["groups"] for room in group["rooms"] if room["room_number"] == room_number)
        points.append(
            {
                "logged_at": now - timedelta(minutes=(23 - index) * 5),
                "average_c": room["average_c"],
                "sensors": room["sensors"],
            }
        )
    return points

File: apps/web/src/api/fixtures.ts
import { expectedSensorTags, roomGroups } from "../config/rooms";
import type { DashboardSnapshot, HistoryPoint, RoomSnapshot, SensorReading, Status } from "../types/dashboard";

function statusFor(value: number | null, quality: SensorReading["quality"]): Status {
  if (quality !== "good" || value === null) return "fault";
  if (value >= 7) return "critical";
  if (value >= 5.5) return "warning";
  return "normal";
}

function roomStatus(statuses: Status[]): Status {
  if (statuses.includes("critical")) return "critical";
  if (statuses.includes("warning")) return "warning";
  if (statuses.includes("fault")) return "fault";
  return "normal";
}

export function fixtureDashboard(): DashboardSnapshot {
  const now = new Date().toISOString();
  const allValues: number[] = [];
  let activeAlerts = 0;
  let warningAlerts = 0;
  let criticalAlerts = 0;
  let sensorFaults = 0;

  const groups = roomGroups.map((group) => {
    const rooms: RoomSnapshot[] = group.roomNumbers.map((roomNumber) => {
      const sensors: SensorReading[] = expectedSensorTags
        .filter((sensor) => sensor.roomNumber === roomNumber)
        .map((sensor, index) => {
          const unavailable = roomNumber === 18;
          const critical = [3, 5, 12, 14].includes(roomNumber);
          const warning = roomNumber === 4 && index === 3;
          const value = unavailable ? null : critical ? 7.1 + index * 0.25 : warning ? 5.8 : 3.4 + ((roomNumber + index) % 5) * 0.45;
          const quality: SensorReading["quality"] = unavailable ? "unavailable" : "good";
          const status = statusFor(value, quality);
          if (status === "fault") sensorFaults += 1;
          if (quality === "good" && value !== null) allValues.push(value);
          return {
            source_tag: sensor.sourceTag,
            label: sensor.label,
            value_c: value === null ? null : Number(value.toFixed(1)),
            quality,
            status
          };
        });

      const values = sensors.flatMap((sensor) => (sensor.value_c === null ? [] : [sensor.value_c]));
      const status = roomStatus(sensors.map((sensor) => sensor.status));
      if (status === "critical") criticalAlerts += 1;
      if (status === "warning") warningAlerts += 1;
      if (status === "critical" || status === "warning") activeAlerts += 1;
      return {
        room_number: roomNumber,
        room_name: `Cold Room ${roomNumber}`,
        group_code: group.groupCode,
        average_c: values.length ? Number((values.reduce((sum, value) => sum + value, 0) / values.length).toFixed(1)) : null,
        status,
        sensors,
        active_alerts: status === "critical" || status === "warning" ? [`${status.toUpperCase()} threshold condition`] : []
      };
    });

    return {
      group_code: group.groupCode,
      room_numbers: [...group.roomNumbers],
      rooms
    };
  });

  return {
    overview: {
      total_rooms: 18,
      total_sensors: 90,
      average_c: Number((allValues.reduce((sum, value) => sum + value, 0) / allValues.length).toFixed(1)),
      active_alerts: activeAlerts,
      critical_alerts: criticalAlerts,
      warning_alerts: warningAlerts,
      sensor_faults: sensorFaults,
      last_log_time: now
    },
    groups
  };
}

export function fixtureHistory(roomNumber: number): HistoryPoint[] {
  return Array.from({ length: 24 }, (_, index) => {
    const snapshot = fixtureDashboard();
    const room = snapshot.groups.flatMap((group) => group.rooms).find((item) => item.room_number === roomNumber);
    return {
      logged_at: new Date(Date.now() - (23 - index) * 300000).toISOString(),
      average_c: room?.average_c ?? null,
      sensors: room?.sensors ?? []
    };
  });
}

